common/src/main/kotlin/streams/service/sink/strategy/CUDIngestionStrategy.kt

```diff
@@ -0,0 +1,263 @@
+            }
+            else -> null
+        }
+    }
+
+    override fun mergeNodeEvents(events: Collection<StreamsSinkEntity>): List<QueryEvents> {
+        val data = events
+                .mapNotNull {
+                    it.value?.let {
+                        try {
+                            val data = toCUDEntity<CUDNode>(it)
+                            when (data?.op)  {
+                                CUDOperations.delete -> null
+                                CUDOperations.merge -> if (data.ids.isNotEmpty()) data else null
+                                else -> data
+                            }
+                        } catch (e: Exception) {
+                            null
+                        }
+                    }
+                }
+                .groupBy({ it.op }, { it })
+
+        val create = data[CUDOperations.create]
+                .orEmpty()
+                .groupBy { it.labels }
+                .map { QueryEvents(buildNodeCreateStatement(it.key), it.value.map { it.toMap() }) }
+        val merge = data[CUDOperations.merge]
+                .orEmpty()
+                .groupBy { it.labels to it.ids.keys }
+                .map { QueryEvents(buildNodeMergeStatement(it.key.first, it.key.second), it.value.map { it.toMap() }) }
+        val update = data[CUDOperations.update]
+                .orEmpty()
+                .groupBy { it.labels to it.ids.keys }
+                .map { QueryEvents(buildNodeUpdateStatement(it.key.first, it.key.second), it.value.map { it.toMap() }) }
+
+        return (create + merge + update) // we'll group the data becase of in case of _id key is present the generated queries are the same for update/merge
+                .map { it.query to it.events }
+                .groupBy({ it.first }, { it.second })
+                .map { QueryEvents(it.key, it.value.flatten()) }
+    }
+
+    override fun deleteNodeEvents(events: Collection<StreamsSinkEntity>): List<QueryEvents> {
+        return events
+                .mapNotNull {
+                    it.value?.let {
+                        try {
+                            val data = toCUDEntity<CUDNode>(it)
+                            when (data?.op)  {
+                                CUDOperations.delete -> if (data.ids.isNotEmpty()) data else null
+                                else -> null
+                            }
+                        } catch (e: Exception) {
+                            null
+                        }
+                    }
+                }
+                .groupBy { Triple(it.labels, it.ids.keys, it.detach) }
+                .map {
+                    val (labels, keys, detach) = it.key
+                    QueryEvents(buildDeleteStatement(labels, keys, detach), it.value.map { it.toMap() })
+                }
+    }
+
+    override fun mergeRelationshipEvents(events: Collection<StreamsSinkEntity>): List<QueryEvents> {
+        val data = events
+                .mapNotNull {
```

Comment:
you have a couple of functions like this which are mapping through all of the events of a certain CUD type.

Same comment on all of them -- you're doing some data validation like the event has to be not null, the operation has to make sense, all of that's good.  But this code is dropping on the floor any kafka message that violates its assumptions.

This is a spot where it'd be good to have a debug level statement saying "Hey, I got some messages but I'm dropping them on the floor because they don't apply or don't make sense to this scenario".   When there's not an option to get this debugging output, the behavior of the plugin may appear buggy or mysterious.

Reply:
Ok I opened an issue for that, as it's a problem common to all strategies.
Look at #237 
Thanks!